File: back-end/src/main/scala/com/idak/tuto/api/config/CustomAuthentProvider.scala
package com.idak.tuto.api.config

import java.util

import com.idak.tuto.api.service.UserService
import org.springframework.security.authentication.{AuthenticationProvider, BadCredentialsException, UsernamePasswordAuthenticationToken}
import org.springframework.security.core.{Authentication, GrantedAuthority}
import org.springframework.stereotype.Component

@Component
class CustomAuthentProvider(userService: UserService) extends AuthenticationProvider {

  @throws[BadCredentialsException]
  override def authenticate(authentication: Authentication) : Authentication = {
    val username = authentication.getName
    val password = authentication.getCredentials.toString
    if (userService.authenticate(username, password).nonEmpty)
      new UsernamePasswordAuthenticationToken(username, password, new util.ArrayList[GrantedAuthority])
    else throw new BadCredentialsException("Invalid Credentials !")
  }

  override def supports(authentication: Class[_]) = authentication.equals(classOf[UsernamePasswordAuthenticationToken])
}

File: back-end-java/src/main/java/com/idak/tuto/api/config/security/CustomAuthentProvider.java
package com.idak.tuto.api.config.security;

import com.idak.tuto.api.service.UserService;
import org.springframework.security.authentication.AuthenticationProvider;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Component;

import javax.annotation.Resource;
import java.util.ArrayList;

@Component
public class CustomAuthentProvider implements AuthenticationProvider {

    @Resource
    private UserService userService;

    @Override
    public Authentication authenticate(Authentication authentication) {
        String username = authentication.getName();
        String password = authentication.getCredentials().toString();

        if(userService.authenticate(username, password).isPresent()){
            return new UsernamePasswordAuthenticationToken(username, password, new ArrayList<>());
        }else{
            throw new BadCredentialsException("Invalid Credentials !");
        }
    }

    @Override
    public boolean supports(Class<?> authentication) {
        return authentication.equals(UsernamePasswordAuthenticationToken.class);
    }
}
